File: app/services/schema_inspector.rb
# frozen_string_literal: true

module Tabami
  module Services
    class SchemaInspector
      def self.extract_identifier_name(ident)
        return "" if ident.nil?

        if defined?(Sequel::SQL::QualifiedIdentifier) && ident.is_a?(Sequel::SQL::QualifiedIdentifier)
          extract_identifier_name(ident.column)
        elsif defined?(Sequel::SQL::Identifier) && ident.is_a?(Sequel::SQL::Identifier)
          extract_identifier_name(ident.value)
        elsif ident.is_a?(Symbol) || ident.is_a?(String)
          ident.to_s
        elsif ident.respond_to?(:column)
          extract_identifier_name(ident.column)
        elsif ident.respond_to?(:value)
          extract_identifier_name(ident.value)
        else
          ident.to_s
        end
      end

      def self.extract_identifier_schema(ident)
        return nil if ident.nil?

        if defined?(Sequel::SQL::QualifiedIdentifier) && ident.is_a?(Sequel::SQL::QualifiedIdentifier)
          extract_identifier_name(ident.table)
        elsif ident.respond_to?(:table)
          extract_identifier_name(ident.table)
        else
          nil
        end
      end

      def self.inspect_schemas(db, adapter)
        case adapter.to_s.downcase
        when "postgres", "postgresql"
          begin
            rows = db.fetch("SELECT schema_name FROM information_schema.schemata WHERE schema_name NOT IN ('pg_toast', 'pg_temp_1', 'pg_toast_temp_1') ORDER BY (schema_name = 'public') DESC, schema_name ASC").all
            schemas = rows.map { |r| (r[:schema_name] || r.values.first).to_s }
            schemas.any? ? schemas : ["public"]
          rescue StandardError
            ["public"]
          end
        when "sqlite"
          ["main"]
        when "mysql", "mysql2"
          begin
            db.fetch("SHOW SCHEMAS").map { |r| r.values.first.to_s }.sort
          rescue StandardError
            ["default"]
          end
        else
          ["public"]
        end
      rescue StandardError
        ["public"]
      end

      def self.inspect_tables(db, adapter, schema = nil)
        tables = []

        case adapter.to_s.downcase
        when "postgres", "postgresql"
          target_schema = (schema.to_s.strip.empty? || schema == "default") ? "public" : schema.to_s.strip

          begin
            sql = <<~SQL
              SELECT 
                c.relname AS table_name,
                CASE c.relkind 
                  WHEN 'r' THEN 'table'
                  WHEN 'v' THEN 'view'
                  WHEN 'm' THEN 'view'
                  WHEN 'f' THEN 'table'
                  WHEN 'p' THEN 'table'
                  ELSE 'table'
                END AS table_type,
                n.nspname AS table_schema,
                c.reltuples::bigint AS estimated_rows
              FROM pg_catalog.pg_class c
              JOIN pg_catalog.pg_namespace n ON n.oid = c.relnamespace
              WHERE n.nspname = ?
                AND c.relkind IN ('r', 'v', 'm', 'f', 'p')
              ORDER BY c.relname;
            SQL

            rows = db.fetch(sql, target_schema).all
            if rows.any?
              tables = rows.map do |r|
                est = r[:estimated_rows]
                tuples = est && est.to_i >= 0 ? est.to_i : nil
                {
                  name: r[:table_name].to_s,
                  type: r[:table_type].to_s.include?("view") ? "view" : "table",
                  schema: target_schema,
                  estimated_rows: tuples
                }
              end
            else
              info_sql = "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = ? ORDER BY table_name"
              tables = db.fetch(info_sql, target_schema).map do |r|
                is_view = r[:table_type].to_s.include?("VIEW")
                {
                  name: r[:table_name].to_s,
                  type: is_view ? "view" : "table",
                  schema: target_schema,
                  estimated_rows: nil
                }
              end
            end
          rescue StandardError
            begin
              info_sql = "SELECT table_name, table_type FROM information_schema.tables WHERE table_schema = ? ORDER BY table_name"
              tables = db.fetch(info_sql, target_schema).map do |r|
                is_view = r[:table_type].to_s.include?("VIEW")
                {
                  name: r[:table_name].to_s,
                  type: is_view ? "view" : "table",
                  schema: target_schema,
                  estimated_rows: nil
                }
              end
            rescue StandardError
              tables = []
            end
          end

        when "sqlite"
          table_names = db.tables.map { |t| extract_identifier_name(t) }.sort
          view_names = db.views.map { |v| extract_identifier_name(v) }.sort

          tables = table_names.map do |t_name|
            count = begin
              db[t_name.to_sym].count
            rescue StandardError
              0
            end
            {
              name: t_name,
              type: "table",
              schema: "main",
              estimated_rows: count
            }
          end

          view_names.each do |v_name|
            tables << {
              name: v_name,
              type: "view",
              schema: "main",
              estimated_rows: nil
            }
          end

        when "mysql", "mysql2"
          db_tables = db.tables.map { |t| extract_identifier_name(t) }.sort
          tables = db_tables.map do |t_name|
            count = begin
              db[t_name.to_sym].count
            rescue StandardError
              0
            end
            {
              name: t_name,
              type: "table",
              schema: db.opts[:database] || "default",
              estimated_rows: count
            }
          end
        end

        tables
      rescue StandardError
        []
      end

      def self.inspect_structure(db, adapter, table_name, schema = nil, reload: false)
        table_sym = schema && adapter.to_s.include?("postgres") ? Sequel[schema.to_sym][table_name.to_sym] : table_name.to_sym

        # 1. Columns schema
        raw_schema = begin
          db.schema(table_sym, reload: reload)
        rescue StandardError
          []
        end

        columns = raw_schema.map do |col_name, info|
          {
            name: extract_identifier_name(col_name),
            type: info[:type].to_s,
            db_type: info[:db_type].to_s,
            allow_null: info[:allow_null] || false,
            default: info[:default]&.to_s,
            primary_key: info[:primary_key] || false,
            max_length: info[:max_length],
            numeric_precision: info[:numeric_precision],
            numeric_scale: info[:numeric_scale]
          }
        end

        # 2. Primary Keys
        pk_columns = columns.select { |c| c[:primary_key] }.map { |c| c[:name] }
        if pk_columns.empty?
          begin
            pk_columns = Array(db.primary_key(table_sym)).map { |c| extract_identifier_name(c) }
          rescue StandardError
            # No primary key found
          end
        end

        # 3. Foreign Keys
        foreign_keys = begin
          db.foreign_key_list(table_sym).map do |fk|
            fk_name = extract_identifier_name(fk[:name])
            cols = Array(fk[:columns]).map { |c| extract_identifier_name(c) }
            {
              name: fk_name.empty? ? "fk_#{cols.join('_')}" : fk_name,
              columns: cols,
              table: extract_identifier_name(fk[:table]),
              schema: extract_identifier_schema(fk[:table]),
              key: Array(fk[:key]).map { |k| extract_identifier_name(k) },
              on_delete: fk[:on_delete]&.to_s,
              on_update: fk[:on_update]&.to_s
            }
          end
        rescue StandardError
          []
        end

        # 4. Indexes
        indexes = begin
          db.indexes(table_sym).map do |idx_name, idx_info|
            {
              name: extract_identifier_name(idx_name),
              columns: Array(idx_info[:columns]).map { |c| extract_identifier_name(c) },
              unique: idx_info[:unique] || false
            }
          end
        rescue StandardError
          []
        end

        # 5. Row count
        total_rows = begin
          db[table_sym].count
        rescue StandardError
          0
        end

        {
          table_name: table_name,
          schema: schema || "public",
          columns: columns,
          primary_keys: pk_columns,
          foreign_keys: foreign_keys,
          indexes: indexes,
          total_rows: total_rows
        }
      end
    end
  end
end

File: src-tauri/src/schema_inspector.rs
use sqlx::Row;
use crate::connection_manager::DbPool;
use crate::types::{ColumnInfo, ForeignKeyInfo, IndexInfo, TableItem, TableStructure};

pub struct SchemaInspector;

impl SchemaInspector {
    pub async fn list_schemas(pool: &DbPool) -> Result<Vec<String>, String> {
        match pool {
            DbPool::Sqlite(_) => Ok(vec!["main".to_string()]),
            DbPool::Postgres(pg) => {
                let rows = sqlx::query(
                    r#"
                    SELECT schema_name 
                    FROM information_schema.schemata 
                    WHERE schema_name NOT IN ('pg_catalog', 'information_schema', 'pg_toast')
                    ORDER BY schema_name
                    "#,
                )
                .fetch_all(pg)
                .await
                .map_err(|e| e.to_string())?;

                let schemas: Vec<String> = rows.into_iter().map(|r| r.get("schema_name")).collect();
                if schemas.is_empty() {
                    Ok(vec!["public".to_string()])
                } else {
                    Ok(schemas)
                }
            }
            DbPool::MySql(my) => {
                let rows = sqlx::query("SELECT SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA WHERE SCHEMA_NAME NOT IN ('information_schema', 'mysql', 'performance_schema', 'sys')")
                    .fetch_all(my)
                    .await
                    .map_err(|e| e.to_string())?;

                let schemas: Vec<String> = rows.into_iter().map(|r| r.get("SCHEMA_NAME")).collect();
                Ok(schemas)
            }
        }
    }

    pub async fn list_tables(pool: &DbPool, schema: Option<&str>) -> Result<Vec<TableItem>, String> {
        match pool {
            DbPool::Sqlite(sqlite) => {
                let rows = sqlx::query(
                    r#"
                    SELECT name, type 
                    FROM sqlite_master 
                    WHERE type IN ('table', 'view') AND name NOT LIKE 'sqlite_%'
                    ORDER BY name
                    "#,
                )
                .fetch_all(sqlite)
                .await
                .map_err(|e| e.to_string())?;

                let mut tables = Vec::new();
                for r in rows {
                    let name: String = r.get("name");
                    let table_type: String = r.get("type");

                    let count: Option<i64> = if table_type == "table" {
                        let count_sql = format!("SELECT COUNT(*) FROM \"{}\"", name.replace('"', "\"\""));
                        sqlx::query_scalar(&count_sql)
                            .fetch_one(sqlite)
                            .await
                            .ok()
                    } else {
                        None
                    };

                    tables.push(TableItem {
                        name,
                        table_type,
                        schema: Some("main".to_string()),
                        estimated_rows: count,
                    });
                }

                Ok(tables)
            }
            DbPool::Postgres(pg) => {
                let target_schema = schema.unwrap_or("public");
                let rows = sqlx::query(
                    r#"
                    SELECT 
                        table_name,
                        table_type,
                        (
                            SELECT COALESCE(c.reltuples, 0)::bigint
                            FROM pg_class c
                            JOIN pg_namespace n ON n.oid = c.relnamespace
                            WHERE n.nspname = table_schema AND c.relname = table_name
                        ) AS estimated_rows
                    FROM information_schema.tables
                    WHERE table_schema = $1
                    ORDER BY table_name
                    "#,
                )
                .bind(target_schema)
                .fetch_all(pg)
                .await
                .map_err(|e| e.to_string())?;

                let mut tables = Vec::new();
                for r in rows {
                    let name: String = r.get("table_name");
                    let raw_type: String = r.get("table_type");
                    let table_type = if raw_type.contains("VIEW") { "view" } else { "table" }.to_string();
                    let estimated: Option<i64> = r.try_get("estimated_rows").ok();

                    tables.push(TableItem {
                        name,
                        table_type,
                        schema: Some(target_schema.to_string()),
                        estimated_rows: estimated,
                    });
                }

                Ok(tables)
            }
            DbPool::MySql(my) => {
                let rows = sqlx::query(
                    r#"
                    SELECT 
                        TABLE_NAME, 
                        TABLE_TYPE, 
                        TABLE_ROWS 
                    FROM INFORMATION_SCHEMA.TABLES 
                    WHERE TABLE_SCHEMA = DATABASE()
                    ORDER BY TABLE_NAME
                    "#,
                )
                .fetch_all(my)
                .await
                .map_err(|e| e.to_string())?;

                let mut tables = Vec::new();
                for r in rows {
                    let name: String = r.get("TABLE_NAME");
                    let raw_type: String = r.get("TABLE_TYPE");
                    let table_type = if raw_type.contains("VIEW") { "view" } else { "table" }.to_string();
                    let rows_count: Option<i64> = r.try_get("TABLE_ROWS").ok();

                    tables.push(TableItem {
                        name,
                        table_type,
                        schema: schema.map(|s| s.to_string()),
                        estimated_rows: rows_count,
                    });
                }

                Ok(tables)
            }
        }
    }

    pub async fn get_table_structure(
        pool: &DbPool,
        table: &str,
        schema: Option<&str>,
    ) -> Result<TableStructure, String> {
        match pool {
            DbPool::Sqlite(sqlite) => {
                let pragma_sql = format!("PRAGMA table_info(\"{}\")", table.replace('"', "\"\""));
                let col_rows = sqlx::query(&pragma_sql)
                    .fetch_all(sqlite)
                    .await
                    .map_err(|e| e.to_string())?;

                let mut columns = Vec::new();
                let mut primary_keys = Vec::new();

                for r in col_rows {
                    let name: String = r.get("name");
                    let db_type: String = r.get("type");
                    let notnull: i32 = r.get("notnull");
                    let dflt_value: Option<String> = r.get("dflt_value");
                    let pk: i32 = r.get("pk");

                    let is_pk = pk > 0;
                    if is_pk {
                        primary_keys.push(name.clone());
                    }

                    let simplified_type = Self::normalize_type(&db_type);

                    columns.push(ColumnInfo {
                        name,
                        col_type: simplified_type,
                        db_type,
                        allow_null: notnull == 0,
                        default: dflt_value,
                        primary_key: is_pk,
                    });
                }

                // Foreign keys
                let fk_sql = format!("PRAGMA foreign_key_list(\"{}\")", table.replace('"', "\"\""));
                let fk_rows = sqlx::query(&fk_sql).fetch_all(sqlite).await.unwrap_or_default();
                let mut foreign_keys = Vec::new();
                for r in fk_rows {
                    let from_col: String = r.get("from");
                    let to_table: String = r.get("table");
                    let to_col: String = r.get("to");
                    let on_delete: Option<String> = r.try_get("on_delete").ok();

                    foreign_keys.push(ForeignKeyInfo {
                        name: format!("fk_{}_{}", table, from_col),
                        columns: vec![from_col],
                        table: to_table,
                        key: vec![to_col],
                        on_delete,
                    });
                }

                // Indexes
                let idx_sql = format!("PRAGMA index_list(\"{}\")", table.replace('"', "\"\""));
                let idx_rows = sqlx::query(&idx_sql).fetch_all(sqlite).await.unwrap_or_default();
                let mut indexes = Vec::new();
                for r in idx_rows {
                    let idx_name: String = r.get("name");
                    let unique: i32 = r.get("unique");

                    let idx_info_sql = format!("PRAGMA index_info(\"{}\")", idx_name.replace('"', "\"\""));
                    let idx_info_rows = sqlx::query(&idx_info_sql).fetch_all(sqlite).await.unwrap_or_default();
                    let idx_cols: Vec<String> = idx_info_rows.into_iter().map(|ir| ir.get("name")).collect();

                    indexes.push(IndexInfo {
                        name: idx_name,
                        columns: idx_cols,
                        unique: unique != 0,
                    });
                }

                let total_rows: i64 = sqlx::query_scalar(&format!("SELECT COUNT(*) FROM \"{}\"", table.replace('"', "\"\"")))
                    .fetch_one(sqlite)
                    .await
                    .unwrap_or(0);

                Ok(TableStructure {
                    table_name: table.to_string(),
                    schema: schema.unwrap_or("main").to_string(),
                    columns,
                    primary_keys,
                    foreign_keys,
                    indexes,
                    total_rows,
                })
            }
            DbPool::Postgres(pg) => {
                let target_schema = schema.unwrap_or("public");

                // Columns
                let col_sql = r#"
                    SELECT 
                        column_name,
                        data_type,
                        udt_name,
                        is_nullable,
                        column_default,
                        EXISTS (
                            SELECT 1 
                            FROM information_schema.table_constraints tc
                            JOIN information_schema.key_column_usage kcu 
                                ON tc.constraint_name = kcu.constraint_name 
                                AND tc.table_schema = kcu.table_schema
                            WHERE tc.constraint_type = 'PRIMARY KEY' 
                                AND tc.table_schema = $1 
                                AND tc.table_name = $2 
                                AND kcu.column_name = c.column_name
                        ) AS is_pk
                    FROM information_schema.columns c
                    WHERE table_schema = $1 AND table_name = $2
                    ORDER BY ordinal_position
                "#;

                let col_rows = sqlx::query(col_sql)
                    .bind(target_schema)
                    .bind(table)
                    .fetch_all(pg)
                    .await
                    .map_err(|e| e.to_string())?;

                let mut columns = Vec::new();
                let mut primary_keys = Vec::new();

                for r in col_rows {
                    let name: String = r.get("column_name");
                    let data_type: String = r.get("data_type");
                    let udt_name: String = r.get("udt_name");
                    let is_nullable: String = r.get("is_nullable");
                    let column_default: Option<String> = r.get("column_default");
                    let is_pk: bool = r.get("is_pk");

                    if is_pk {
                        primary_keys.push(name.clone());
                    }

                    let simplified_type = Self::normalize_type(&udt_name);

                    columns.push(ColumnInfo {
                        name,
                        col_type: simplified_type,
                        db_type: data_type,
                        allow_null: is_nullable == "YES",
                        default: column_default,
                        primary_key: is_pk,
                    });
                }

                let total_rows: i64 = sqlx::query_scalar(&format!("SELECT COUNT(*) FROM \"{}\".\"{}\"", target_schema.replace('"', "\"\""), table.replace('"', "\"\"")))
                    .fetch_one(pg)
                    .await
                    .unwrap_or(0);

                Ok(TableStructure {
                    table_name: table.to_string(),
                    schema: target_schema.to_string(),
                    columns,
                    primary_keys,
                    foreign_keys: Vec::new(),
                    indexes: Vec::new(),
                    total_rows,
                })
            }
            DbPool::MySql(my) => {
                let col_rows = sqlx::query(
                    r#"
                    SELECT 
                        COLUMN_NAME, 
                        DATA_TYPE, 
                        COLUMN_TYPE, 
                        IS_NULLABLE, 
                        COLUMN_DEFAULT, 
                        COLUMN_KEY 
                    FROM INFORMATION_SCHEMA.COLUMNS 
                    WHERE TABLE_SCHEMA = DATABASE() AND TABLE_NAME = ?
                    ORDER BY ORDINAL_POSITION
                    "#,
                )
                .bind(table)
                .fetch_all(my)
                .await
                .map_err(|e| e.to_string())?;

                let mut columns = Vec::new();
                let mut primary_keys = Vec::new();

                for r in col_rows {
                    let name: String = r.get("COLUMN_NAME");
                    let data_type: String = r.get("DATA_TYPE");
                    let col_type: String = r.get("COLUMN_TYPE");
                    let is_nullable: String = r.get("IS_NULLABLE");
                    let dflt: Option<String> = r.get("COLUMN_DEFAULT");
                    let key: String = r.get("COLUMN_KEY");

                    let is_pk = key == "PRI";
                    if is_pk {
                        primary_keys.push(name.clone());
                    }

                    let simplified_type = Self::normalize_type(&data_type);

                    columns.push(ColumnInfo {
                        name,
                        col_type: simplified_type,
                        db_type: col_type,
                        allow_null: is_nullable == "YES",
                        default: dflt,
                        primary_key: is_pk,
                    });
                }

                let total_rows: i64 = sqlx::query_scalar(&format!("SELECT COUNT(*) FROM `{}`", table.replace('`', "``")))
                    .fetch_one(my)
                    .await
                    .unwrap_or(0);

                Ok(TableStructure {
                    table_name: table.to_string(),
                    schema: schema.unwrap_or("").to_string(),
                    columns,
                    primary_keys,
                    foreign_keys: Vec::new(),
                    indexes: Vec::new(),
                    total_rows,
                })
            }
        }
    }

    fn normalize_type(db_type: &str) -> String {
        let t = db_type.to_lowercase();
        if t.contains("int") {
            "integer".to_string()
        } else if t.contains("char") || t.contains("text") || t.contains("clob") {
            "string".to_string()
        } else if t.contains("float") || t.contains("double") || t.contains("decimal") || t.contains("numeric") || t.contains("real") {
            "float".to_string()
        } else if t.contains("bool") {
            "boolean".to_string()
        } else if t.contains("date") || t.contains("time") {
            "datetime".to_string()
        } else if t.contains("json") {
            "json".to_string()
        } else if t.contains("blob") || t.contains("bytea") {
            "binary".to_string()
        } else {
            "string".to_string()
        }
    }
}
